modules/core/src/mathfuncs.cpp

```diff
@@ -1600,50 +1608,311 @@ static bool ocl_patchNaNs( InputOutputArray _a, float value )
+}
+
+template <>
+int finiteMaskSIMD_<double, 1>(const double *src, uchar *dst, size_t total)
+{
+    const int size8 = v_uint8::nlanes;
+    int i = 0;
+    for(; i <= (int)total - (size8 / 2); i += (size8 / 2) )
+    {
+        v_uint64 vu[4];
+        for (int j = 0; j < 4; j++)
+            vu[j] = vx_load((const uint64_t*)src + i + j*(size8 / 8));
+
+        v_uint64 vmaskExp = vx_setall_u64(0x7ff0000000000000);
+        v_uint64 z = vx_setzero_u64();
+        #if !CV_SIMD128_64F
+        v_uint64 mask10 = vx_setall_u64(0xffffffff00000000);
+        #endif
+        v_uint64 vv[4];
+        for (int j = 0; j < 4; j++)
+        {
+            v_uint64 vande = vu[j] & vmaskExp;
+
+            v_uint64 ve;
+            #if CV_SIMD128_64F
+            ve = vande != vmaskExp;
+            #else
+            // emulating 64-bit integer eq comparison: ve = (vu[j] & vmaskExp) == vmaskExp
+            {
+                v_int32 vue32 = v_reinterpret_as_s32(vande);
+                v_int32 vme32 = v_reinterpret_as_s32(vmaskExp);
+                v_int32 veq32 = vue32 != vme32;
+                v_int32 sh1 = v_rotate_left<1>(veq32);
+                v_int32 vand = veq32 | sh1;
+                v_int32 sh2 = v_rotate_right<1>(vand);
+                ve = (v_reinterpret_as_u64(vand) & mask10) | (v_reinterpret_as_u64(sh2) & ~mask10);
+            }
+            #endif
+
+            vv[j] = ve;
+        }
+
+        v_uint8 v = v_pack_b(vv[0], vv[1], vv[2], vv[3], z, z, z, z);
+
+        v_store_low(dst + i, v);
+    }
+
+    return i;
+}
+
+#endif
+
+
+template <typename _Tp, int cn>
+static void finiteMask_(const _Tp *src, uchar *dst, size_t total)
+{
+    size_t i = 0;
+
+#if CV_SIMD
+    i = finiteMaskSIMD_<_Tp, cn>(src, dst, total);
 #endif
 
-        for( ; j < len; j++ )
-            if( (tptr[j] & 0x7fffffff) > 0x7f800000 )
-                tptr[j] = val.i;
+    for(; i < total; i++ )
+    {
+        bool nan = true;
+        for (int c = 0; c < cn; c++)
+        {
+            _Tp val = src[i * cn + c];
+            nan = nan && !cvIsNaN(val) && !cvIsInf(val);
+        }
+        dst[i] = nan ? 255 : 0;
     }
 }
 
+#ifdef HAVE_OPENCL
+
+
+static bool ocl_finiteMask(const UMat img, UMat mask)
+{
+    int channels = img.channels();
+    int depth = img.depth();
+    int rowsPerWI = ocl::Device::getDefault().isIntel() ? 4 : 1;
+    ocl::Kernel k("finiteMask", ocl::core::finitemask_oclsrc,
+                  format("-D srcT=%s -D cn=%d -D rowsPerWI=%d",
+                         depth == CV_32F ? "float" : "double", channels, rowsPerWI));
+    if (k.empty())
+        return false;
+
+    k.args(ocl::KernelArg::ReadOnlyNoSize(img), ocl::KernelArg::WriteOnly(mask));
+
+    size_t globalsize[2] = { (size_t)img.cols, ((size_t)img.rows + rowsPerWI - 1) / rowsPerWI };
+    return k.run(2, globalsize, NULL, false);
 }
 
+#endif
+
+void finiteMask(InputArray _img, OutputArray _mask)
+{
+    CV_INSTRUMENT_REGION();
+
+    int channels = _img.channels();
```

Comment:
`channels=5` doesn't throw any exception and do nothing.

Reply:
Fixed